### PR TITLE
Adding features to ilo_redfish_utils.py

### PR DESCRIPTION
**Pull request by TSKushal**
SUMMARY
Adding features to ilo_redfish_utils.py
The modules for these functions will be raised in a separate PR for ease of merging.
This was originally raised in [5328](https://github.com/ansible-collections/community.general/pull/5328)

ISSUE TYPE
- Feature Pull Request

COMPONENT NAME
ilo_redfish_utils.py

### CODE REVIEW

**Comment by TSKushal** [comment]
@ansibullbot  /rebuild_failed

**Comment by TSKushal** [comment]
cc @Bhavya06 @mraineri @rajeevkallur @renxulei @tomasg2012 @xmadsen

File: plugins/module_utils/ilo_redfish_utils.py [hunk 1]
```diff
@@ -51,6 +99,423 @@ def get_ilo_sessions(self):
         result["ret"] = True
         return result
 
+    def get_systems_data(self):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:102]: Getting system info is standard Redfish functionality; we have this today as "get_system_inventory". Why is this needed?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 2]
```diff
@@ -51,6 +99,423 @@ def get_ilo_sessions(self):
...
+        response = self.get_request(self.root_uri + self.systems_uri)
+        if not response["ret"]:
+            return response
+
+        return {
+            "ret": True,
+            "msg": response
+        }
+
+    def get_managers_data(self):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:112]: Getting manager info is standard Redfish functionality; we have this today as "get_manager_inventory". Why is this needed in an iLO-specific module?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 3]
```diff
@@ -51,6 +99,423 @@ def get_ilo_sessions(self):
...
+        response = self.get_request(self.root_uri + self.manager_uri)
+        if not response["ret"]:
+            return response
+
+        return {
+            "ret": True,
+            "msg": response
+        }
+
+    def get_service_bios_attributes(self):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:122]: Getting BIOS attributes is standard Redfish functionality; we have this today as "get_bios_attributes". Why is this needed in an iLO-specific module?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 4]
```diff
@@ -51,6 +99,423 @@ def get_ilo_sessions(self):
...
+            result["msg"] = self.remove_odata(response["data"])
+        else:
+            return {
+                "ret": False,
+                "msg": "Boot settings uri not found in %s response, %s" % (uri, data)
+            }
+
+        return result
+
+    def get_physical_drives_details(self):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:184]: Does iLO not support standard drive resources? This functionality is all covered by standard Redfish, but it's very odd to go into OEM space for this type of info.

File: plugins/module_utils/ilo_redfish_utils.py [hunk 5]
```diff
@@ -231,3 +696,1561 @@ def set_wins_registration(self, mgrattr):
...
+                "ret": False,
+                "changed": False,
+                "msg": "Logical drive count in raid details is not matching with logical drives present in the server"
+            }
+        return {
+            "ret": True,
+            "msg": "Drive count is same as input"
+        }
+
+    def set_service_bios_attributes(self, service_attributes):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:750]: Getting BIOS attributes is standard Redfish functionality; we have this today as "set_bios_attributes". Why is this needed in an iLO-specific module?

File: plugins/module_utils/ilo_redfish_utils.py (hunk 1 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:102]: This function is being used in every subsequent function in the file to retrieve the entire systems data. **"get_system_inventory"** returns only certain attributes, whereas we need many different attributes in many of the function.
Please refer get_service_bios_attributes() for this.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 3 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:122]: get_service_bios_attributes() queries a different a API then the get_bios_attributes() function
get_service_bios_attributes API - oem/hpe/service/settings/ - This provides the service options available in the server.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 5 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:750]: This is also a different API. Please refer the above comment.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 1 above)
Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:102]: I would much rather enhance "get_system_inventory" in the standard module (or even have a different method in the standard module) to retrieve all properties in a desired system. This is not something iLO-specific.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 2 above)
Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:112]: I would much rather enhance "get_manager_inventory" in the standard module (or even have a different method in the standard module) to retrieve all properties in a desired manager. This is not something iLO-specific.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 3 above)
Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:122]: I see that now; as long as this goes off into OEM, then I agree this can stay here.

File: plugins/module_utils/ilo_redfish_utils.py [hunk 6]
```diff
@@ -51,6 +99,423 @@ def get_ilo_sessions(self):
...
+                       % (self.root_uri + self.service_root, str(data))
+            }
+
+        bios_uri = data["Bios"]["@odata.id"]
+
+        # Get service settings
+        service_uri = bios_uri + "service/settings/"
+        response = self.get_request(self.root_uri + service_uri)
+        # Check if service API doesn't support
+        if not response["ret"]:
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:144]: Looking at this logic a bit more, it seems to only go into the OEM path if the standard path failed. Is this a realistic scenario?

Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:144]: Nevermind; `bios_uri + "service/settings/"` isn't a standard URI path, but it also doesn't contain `Oem` like the Redfish spec defines for OEM URIs. What exactly is found at `bios_uri + "service/settings/"`?

File: plugins/module_utils/ilo_redfish_utils.py (hunk 3 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:122]: Thank you, closing this conversation thread.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 6 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:144]: bios_uri + "service/settings/" - is for Gen10 server model
bios_uri + "oem/hpe/service/settings/" - is for Gen10 Plus server model 
Both will fetch the service BIOS attributes.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 1 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:102]: Hi, we noticed that this seems to be an unnecessary function. An oversight on our part, referred **redfish_utils** and will be following that as an example. Sorry for the confusion.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 2 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:112]: Hi, Please refer the above comment, the same is applicable here.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 4 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:164]: Hi, i think we discussed this in the other PR [5540](https://github.com/ansible-collections/community.general/pull/5540), please refer this.

Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:164]: Could we perhaps rename this method (and others touching "SmartStorage" to specifically have "SmartStorage" in the method name (such as get_smartstorage_physical_drives_details) to avoid confusion with getting drives in a generic manner?

Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Changed get_physical_drives_details() to get_smartstorage_physical_drives().

File: plugins/module_utils/ilo_redfish_utils.py [hunk 7]
```diff
@@ -51,6 +99,403 @@ def get_ilo_sessions(self):
...
+            result["physical_drives"] = physical_drives
+            result["physical_drives_count"] = physical_drives_count
+
+        return {
+            "msg": result,
+            "ret": True
+        }
+        # return result
+
+    def get_logical_drives_details(self, array_controllers=False):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:242]: Can we make the same change here (and other places where "SmartStorage" is the target Storage subsystem) to include "smartstorage" in the method name?

Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Updated "get_logical_drives_details" to "get_smartstorage_logical_drives". Keeping other function names such as "create_logical_drives" the same as they don't have an equivalent with storage or simaplstorage in redfish_utils

File: plugins/module_utils/ilo_redfish_utils.py (hunk 1 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 2 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 4 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 5 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Removed this functionality now. Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 6 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Removed this functionality for now. Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 7 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation, thanks!

File: plugins/module_utils/ilo_redfish_utils.py [hunk 8]
```diff
@@ -51,6 +99,335 @@ def get_ilo_sessions(self):
...
+                    return response
+                output["SNMPv3User"] = self.remove_odata(response["data"])
+            alert_destinations.append(output)
+
+        return {
+            "ret": True,
+            "msg": alert_destinations
+        }
+
+    def verify_bios_attributes(self, bios_attributes):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:389]: Could this be moved to the standard Redfish module? There's nothing here that looks iLO-specific to me.

File: plugins/module_utils/ilo_redfish_utils.py [hunk 9]
```diff
@@ -231,3 +608,1407 @@ def set_wins_registration(self, mgrattr):
         if not response['ret']:
             return response
         return {'ret': True, 'changed': True, 'msg': "Modified %s" % mgrattr['mgr_attr_name']}
+
+    def verify_drive_count(self, raid_details, logical_drives_count):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:612]: Is this going to remain specific to Smart Storage?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 10]
```diff
@@ -231,3 +608,1407 @@ def set_wins_registration(self, mgrattr):
...
+                "ret": False,
+                "changed": False,
+                "msg": "Logical drive count in raid details is not matching with logical drives present in the server"
+            }
+        return {
+            "ret": True,
+            "msg": "Drive count is same as input"
+        }
+
+    def verify_logical_drives(self, raid_details, check_length=False):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:624]: Is this going to remain specific to Smart Storage?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 11]
```diff
@@ -231,3 +608,1407 @@ def set_wins_registration(self, mgrattr):
...
+        # This method verifies logical drives present in the OOB controller against the provided input
+        result = self.get_logical_drives_details()
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:626]: I think this needs to be updated to use the new method name.

File: plugins/module_utils/ilo_redfish_utils.py [hunk 12]
```diff
@@ -231,3 +608,1407 @@ def set_wins_registration(self, mgrattr):
...
+                    "changed": False,
+                    "msg": "Input Boot Order %s doesn't match with Server Boot Order %s" % (str(input_boot_order), str(server_boot_order))
+                }
+        return {
+            "ret": True,
+            "changed": False,
+            "msg": "Input Boot Order matches with the Server Boot Order"
+        }
+
+    def delete_all_logical_drives(self):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:733]: Is this also Smart Storage-specific?

File: plugins/module_utils/ilo_redfish_utils.py [hunk 13]
```diff
@@ -231,3 +608,1407 @@ def set_wins_registration(self, mgrattr):
...
+                not_defined = []
+                for input in input_list:
+                    if input not in raid.keys():
+                        missing_param.append(input)
+                    elif not raid[input]:
+                        not_defined.append(input)
+
+            return self.validation_error(raid, input_list, missing_param, not_defined, [], command="CreateLogicalDrivesWithParticularPhysicalDrives")
+
+    def check_physical_drives(self, raid_data, unused_physical_drives, command="CreateLogicalDrives"):
```
Comment (mraineri) [inline plugins/module_utils/ilo_redfish_utils.py:938]: Can this be renamed for Smart Storage to avoid confusion with generic drive checking?

File: plugins/module_utils/ilo_redfish_utils.py (hunk 11 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:626]: Yes, updated. Thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 8 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:389]: Hi, 
Yes you're right. Removing it from here and will raise a separate PR to merge this to redfish_command module. Thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 9 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:578]: This function is used internally by the verify_logical_drives functionality. So, it is SmartStorage specific.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 10 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:590]: Yes, this retrieves information about logical drives from the SmartStorage space.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 12 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:699]: Yes, this is SmartStorage specific.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 13 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py:904]: We've renamed all the commands in the other PRs raised to include SmartStorage in the name. 
Ex: GetLogicalDrives to GetSmartStorageLogicalDrives. [5540](https://github.com/ansible-collections/community.general/pull/5540/)
Do we need to rename all the functions being called as well?

Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:904]: Yes, since they're functions of the module, and other developers might be misled as to what their intent is.

Comment (mraineri) [reply plugins/module_utils/ilo_redfish_utils.py:904]: The other concern is if we grow the standard module to include more storage-related functions, this name could conflict with what we'd like to do there.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 8 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation. Thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 11 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation. Thanks!

**Comment by mraineri** [comment]
From a coding perspective everything is addressed on my end.

File: plugins/module_utils/ilo_redfish_utils.py [hunk 14]
```diff
@@ -5,13 +5,60 @@
...
 class iLORedfishUtils(RedfishUtils):
 
+    def __init__(self, creds, root_uri, timeout, module):
+        super().__init__(creds, root_uri, timeout, module)
+
+        if not HAS_IPADDRESS:
+            self.module.fail_json(msg=missing_required_lib('ipaddress'), exception=IPADDRESS_IMP_ERR)
+
+        if not HAS_URLLIB3:
+            self.module.fail_json(msg=missing_required_lib('urllib3'), exception=URLLIB3_IMP_ERR)
```
Comment (felixfontein) [inline plugins/module_utils/ilo_redfish_utils.py:60]: These are breaking changes. The modules using this util worked so far when `ipaddress` and/or `urllib3` were not available, and now they will suddenly fail.

They should only fail when the user requests new features that need these libraries (when they are not around), but it should not fail for cases where the modules already worked in the past when these libraries were not available.

File: changelogs/fragments/5539-adding-features-to-ilo_redfish_utils.yml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - added new features for ilo_redfish_utils (https://github.com/ansible-collections/community.general/pull/5539/).
```
Comment (felixfontein) [inline changelogs/fragments/5539-adding-features-to-ilo_redfish_utils.yml:2]:
```suggestion
  - ilo_redfish_utils module utils - added new features (https://github.com/ansible-collections/community.general/pull/5539/).
```
Would probably be nice to be a little bit more concrete, about what these features are for.

File: plugins/module_utils/ilo_redfish_utils.py (hunk 10 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation. Thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 13 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: All functions have now been renamed. Resolving this conversation. Thanks!

File: plugins/module_utils/ilo_redfish_utils.py (hunk 12 above)
Comment (TSKushal) [reply plugins/module_utils/ilo_redfish_utils.py]: Resolving this conversation. Thanks!

**Comment by TSKushal** [comment]
Closing this PR as we will be pushing these modules into ansible-galaxy collection **"hpe.ilo"**. Thanks to all the reviewers!
